### PR TITLE
Define Headers to Parse during Page Prep - resolve #984

### LINKED ISSUE #984: Add Ability to specify how deep page headers should go
## Feature request
It would be nice to have the ability to specify how deep the `extractHeaders` method goes.
Currently, it has a hard coded value of `['h2', 'h3']`
Our current markdown files use `h4` as well in order to handle some table of contents for the page in sidebar form.

#### What problem does this feature solve?
When doing a table of contents in the sidebar, this feature would allow you to specify how deep the TOC should go based on headers.

#### What does the proposed API look like?
config.js
```
module.exports = {
    themeConfig: {
        parseHeaders: ['h2','h3','h4']
    }
}
```

#### How should this be implemented in your opinion?
The above would be a themeConfig option and will replace the default hard coded value located at 
https://github.com/vuejs/vuepress/blob/master/packages/%40vuepress/core/lib/prepare/Page.js#L92

#### Are you willing to work on this yourself?**
Yes

### PR DESCRIPTION
**Pull request by bretterer**
**Summary**
This PR resolves #984 

Currently the header parsing is hard coded to `['h2', 'h3']`. This code changes that to give the ability to configure this in the `themeOptions` config.

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [x] Docs
- [ ] Build-related changes
- [ ] Other, please describe:

If changing the UI of default theme, please provide the **before/after** screenshot:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)

You have tested in the following browsers: (Providing a detailed version will be better.)

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge
- [ ] IE

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature
- [ ] Related documents have been updated
- [ ] Related tests have been updated

To avoid wasting your time, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

**Comment by youxfans** [comment]
@ulivz  Look here.

**Comment by bretterer** [comment]
Updated.  Thanks!

**Comment by bretterer** [comment]
No worries.  That makes sense.  I will update this tonight/tomorrow and get some updates to you

**Comment by bretterer** [comment]
Sorry, Holidays got in the way... Update coming

**Comment by bretterer** [comment]
@ulivz updated.  

I had to gain access to the config, and I think this is the best way to do it.  Maybe all of this should be extracted out to a `getConfig` module that can be reused?

**Comment by bretterer** [comment]
@ulivz wondered if you had a chance to take a look at this.  We are getting closer to needing this feature in our development.

**Comment by bretterer** [comment]
Would it be possible to get this into `1.0.0-alpha.28` release?

File: packages/@vuepress/markdown-loader/index.js [hunk 1]
```diff
@@ -26,7 +30,10 @@ module.exports = function (src) {
   if (!markdown) {
     markdown = md()
   }
-
+  let config = loadConfig(path.resolve(sourceDir, '.vuepress'), false)
+  if (isFunction(config)) {
+    config = config(this)
+  }
```
Comment (ulivz) [inline packages/@vuepress/markdown-loader/index.js]: You shouldn't re-load user's config at the scope makrdown-loader, this loader should only focus on how to transpile the source markdown file to Vue component.

`extractHeaders` should be a option for this loader.

Comment (bretterer) [reply packages/@vuepress/markdown-loader/index.js]: Ill do some more digging, but I couldn't find any way to get to the `extractHeaders` option from within this file already.

Comment (bretterer) [reply packages/@vuepress/markdown-loader/index.js]: Are you saying it should be available already inside of the loader file if defined in the `.vuepress/config.js` file such as:
```
module.exports = ctx => ({
  markdown: {
    extractHeaders: ['h2','h3','h4','h5']
  }
})
```

if that is the case, I have looked at the `this` object and could not find it anywhere.

**Comment by ulivz** [comment]
To avoid wasting your time, I recommend that this feature be implemented by official collaborators.

**Comment by bretterer** [comment]
@ulivz: And, how would one (Okta and Myself) become an official collaborator for VuePress on an open source tool?  I am pretty sure this was not wasting my time as I was trying to donate my time to help expand this project.

**Comment by knyttl** [comment]
So is this is abandoned and not implemented?

**Comment by bretterer** [comment]
@knyttl this was handled through https://github.com/vuejs/vuepress/pull/1945

**Pull request closed without merging**